src/lines-tile-component.coffee

```diff
@@ -114,64 +114,70 @@ class LinesTileComponent
 
     return
 
-  buildLineHTML: (id) ->
+  buildLineNode: (id) ->
     {width} = @newState
     {screenRow, tokens, text, top, lineEnding, fold, isSoftWrapped, indentLevel, decorationClasses} = @newTileState.lines[id]
 
-    classes = ''
+    lineNode = @elementsPool.build("div", "line")
     if decorationClasses?
       for decorationClass in decorationClasses
-        classes += decorationClass + ' '
-    classes += 'line'
+        lineNode.classList.add(decorationClass)
 
-    lineHTML = "<div class=\"#{classes}\" style=\"position: absolute; top: #{top}px; width: #{width}px;\" data-screen-row=\"#{screenRow}\">"
+    lineNode.style.position = "absolute"
+    lineNode.style.top = top + "px"
+    lineNode.style.width = width + "px"
+    lineNode.dataset.screenRow = screenRow
 
     if text is ""
-      lineHTML += @buildEmptyLineInnerHTML(id)
+      @appendEmptyLineInnerNodes(id, lineNode)
     else
-      lineHTML += @buildLineInnerHTML(id)
+      @appendLineInnerNodes(id, lineNode)
 
-    lineHTML += '<span class="fold-marker"></span>' if fold
-    lineHTML += "</div>"
-    lineHTML
+    lineNode.appendChild(@elementsPool.build("span", "fold-marker")) if fold
+    lineNode
 
-  buildEmptyLineInnerHTML: (id) ->
+  appendEmptyLineInnerNodes: (id, lineNode) ->
     {indentGuidesVisible} = @newState
     {indentLevel, tabLength, endOfLineInvisibles} = @newTileState.lines[id]
 
     if indentGuidesVisible and indentLevel > 0
       invisibleIndex = 0
-      lineHTML = ''
       for i in [0...indentLevel]
-        lineHTML += "<span class='indent-guide'>"
+        indentGuide = @elementsPool.build("span", "indent-guide")
         for j in [0...tabLength]
           if invisible = endOfLineInvisibles?[invisibleIndex++]
-            lineHTML += "<span class='invisible-character'>#{invisible}</span>"
+            indentGuide.appendChild(
+              @elementsPool.build("span", "invisible-character", invisible)
+            )
           else
-            lineHTML += ' '
-        lineHTML += "</span>"
+            indentGuide.insertAdjacentText("beforeend", " ")
+        lineNode.appendChild(indentGuide)
 
       while invisibleIndex < endOfLineInvisibles?.length
-        lineHTML += "<span class='invisible-character'>#{endOfLineInvisibles[invisibleIndex++]}</span>"
-
-      lineHTML
+        invisible = endOfLineInvisibles[invisibleIndex++]
+        lineNode.appendChild(
+          @elementsPool.build("span", "invisible-character", invisible)
+        )
     else
-      @buildEndOfLineHTML(id) or '&nbsp;'
+      unless @appendEndOfLineNodes(id, lineNode)
+        lineNode.insertAdjacentText("beforeend", "\u00a0")
```

Comment:
I think we only need to insert a non-breaking space on empty lines... or am I missing something. That might allow you to avoid this weird `insertAdjacentText` method. Don't care that much though.